### PR TITLE
`Development`: Improve course overview component

### PR DESCRIPTION
**Pull request by rabeatwork**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Enhance the Menu Sidebar Code to facilitate the creation of a reusable generic component for the course management view.

### Description

- Improved the use of NgbDropdown Component (Is used for the `More` Item, which appears on small screens)
- Implemented `ng-template` for reusable code.
- Streamlined and simplified some logic.

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

## Summary by CodeRabbit

- **New Features**
  - Enhanced dropdown functionality in the Course Overview with improved visibility handling.

- **Bug Fixes**
  - Fixed dropdown behavior to ensure consistent open/close actions.

- **Refactor**
  - Renamed methods for better clarity and alignment with new dropdown functionality.
  - Updated test cases to incorporate new dropdown behavior controls.

- **Tests**
  - Implemented tests for new dropdown open/close methods to ensure reliability.

### CODE REVIEW

**Review by JohannesWt — DISMISSED** [review]
Code 👍 (nice refactoring)

**Review by egekurt123 — DISMISSED** [review]
Code LGTM

**Review by florian-glombik — APPROVED** [review]
Tested on ts3, course switching and `More` dropdown worked as expected.

The Code looks good, thanks for the refactoring!

When testing I noticed a flickering of the icon (seems to be present on previous versions, not introduced by this PR - I think this is something we should address.

flickeringOfCourseIcon

**Review by edkaya — APPROVED** [review]
Tested on ts3, didn't face any problems in more option and course switching. Code also lgtm

File: src/test/javascript/spec/helpers/mocks/directive/ngbDropdownMocks.module.ts [hunk 1]
```diff
@@ -28,12 +28,18 @@
...
-    @Input() open = false;
+    @Input() isOpen = false;
     @Input() placement: any;
     @Input() popperOptions: (options: Partial<any>) => Partial<any>;
     @Input() container: null | 'body';
     @Input() display: 'dynamic' | 'static';
     @Output() openChange = new EventEmitter<boolean>();
+    open() {
+        this.isOpen = true;
+    }
```
[the remark this answers is not in the document]
Comment (MaximilianAnzinger) [reply src/test/javascript/spec/helpers/mocks/directive/ngbDropdownMocks.module.ts:39]: I think the usually stupid rabbit is right in this case.

File: src/test/javascript/spec/helpers/mocks/directive/ngbDropdownMocks.module.ts [hunk 2]
```diff
@@ -28,12 +28,18 @@ class NgbDropdownToggleMockDirective {}
...
+    close() {
+        this.isOpen = false;
+    }
```
[the remark this answers is not in the document]
Comment (MaximilianAnzinger) [reply src/test/javascript/spec/helpers/mocks/directive/ngbDropdownMocks.module.ts:42]: same here

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]

**Review by MaximilianAnzinger — DISMISSED** [review]

**Review by edkaya — DISMISSED** [review]
Reapprove after latest changes

**Review by florian-glombik — APPROVED** [review]
Re-approve

**Review by egekurt123 — APPROVED** [review]
Reapprove

**Review by JohannesWt — APPROVED** [review]
Reapprove Code 👍

**Comment by rabeatwork** [comment]
> There hasn't been any activity on this pull request recently. Therefore, this pull request has been automatically marked as stale and will be closed if no further activity occurs within **seven** days. Thank you for your contributions.

Don't close it. :)

**Review by edkaya — APPROVED** [review]
Reapprove

**Review by florian-glombik — CHANGES_REQUESTED** [review]
Tested on ts1, I think that the more tab should close on selection (would have expected the more section to close on selection and it appears to be the behaviour on production right now)

https://github.com/user-attachments/assets/6c1886fb-7e95-4f8f-bb40-eb0df43bcf64

**Comment by rabeatwork** [comment]
> Tested on ts1, I think that the more tab should close on selection (would have expected the more section to close on selection and it appears to be the behaviour on production right now)
> 
>  moreDoesNotCloseOnSelection.mp4

Please deploy the current version again. This should be fixed already. 👯

**Review by florian-glombik — APPROVED** [review]
Tested on ts1, course switching and more button worked fine for me

The mentioned flickering seems out of scope for this PR https://github.com/ls1intum/Artemis/pull/8971#pullrequestreview-2159563686

**Review by JohannesWt — APPROVED** [review]
Code reapprove 👍

**Review by egekurt123 — APPROVED** [review]
Reapprove, looks good on TS1

**Review by MaximilianAnzinger — APPROVED** [review]

**Pull request merged**